File: engines/v8/ts/promise.ts
import type { NativeV8Bindings } from "./engine"
import type { V8PromiseState, V8Result } from "./types"
import { V8Value } from "./value"

export class V8Promise {
  constructor(
    private native: NativeV8Bindings,
    private context: number,
    private handle: number,
  ) {}

  static create(native: NativeV8Bindings, context: number): V8Promise {
    const handle = native.promiseNew(context)
    return new V8Promise(native, context, handle)
  }

  static wrap(native: NativeV8Bindings, context: number, handle: number): V8Promise {
    return new V8Promise(native, context, handle)
  }

  get id(): number {
    return this.handle
  }

  resolve(valueHandle?: number): V8Result<null> {
    return this.native.promiseResolve(this.context, this.handle, valueHandle)
  }

  reject(reason: string): V8Result<null> {
    return this.native.promiseReject(this.context, this.handle, reason)
  }

  get state(): V8PromiseState {
    return this.native.promiseState(this.context, this.handle)
  }

  hasHandler(): boolean {
    return this.native.promiseHasHandler(this.context, this.handle)
  }

  markAsHandled(): V8Result<null> {
    return this.native.promiseMarkAsHandled(this.context, this.handle)
  }

  getNative(): V8Value {
    const valueHandle = this.native.promiseGetNative(this.handle)
    return V8Value.wrap(this.native, this.context, valueHandle)
  }
}

File: engines/v8/src/promise.rs
use crate::error::V8Error;

#[cfg(feature = "native")]
use crate::ffi;

pub struct V8Promise {
    #[cfg(feature = "native")]
    handle: *mut ffi::V8PromiseHandle,
}

impl V8Promise {
    #[cfg(feature = "native")]
    pub fn new(ctx: *mut ffi::V8ContextHandle) -> Result<Self, V8Error> {
        let ptr = unsafe { ffi::klyron_v8_promise_new(ctx) };
        if ptr.is_null() { Err(V8Error::CallFailed("promise creation failed".into())) }
        else { Ok(Self { handle: ptr }) }
    }

    #[cfg(not(feature = "native"))]
    pub fn new(_ctx: *mut std::ffi::c_void) -> Result<Self, V8Error> {
        Err(V8Error::NotInitialized)
    }

    #[cfg(feature = "native")]
    pub fn resolve(&self, ctx: *mut ffi::V8ContextHandle, value: *mut ffi::V8ValueHandle) -> Result<(), V8Error> {
        let r = unsafe { ffi::klyron_v8_promise_resolve(ctx, self.handle, value) };
        if r.success { Ok(()) }
        else { Err(V8Error::CallFailed("promise resolve failed".into())) }
    }

    #[cfg(feature = "native")]
    pub fn reject(&self, ctx: *mut ffi::V8ContextHandle, reason: &str) -> Result<(), V8Error> {
        use std::ffi::CString;
        let c = CString::new(reason).map_err(|e| V8Error::CallFailed(e.to_string()))?;
        let r = unsafe { ffi::klyron_v8_promise_reject(ctx, self.handle, c.as_ptr()) };
        if r.success { Ok(()) }
        else { Err(V8Error::CallFailed("promise reject failed".into())) }
    }

    #[cfg(feature = "native")]
    pub fn state(&self, ctx: *mut ffi::V8ContextHandle) -> crate::V8PromiseState {
        let s = unsafe { ffi::klyron_v8_promise_get_state(ctx, self.handle) };
        match s {
            1 => crate::V8PromiseState::Fulfilled,
            2 => crate::V8PromiseState::Rejected,
            _ => crate::V8PromiseState::Pending,
        }
    }

    #[cfg(feature = "native")]
    pub fn has_handler(&self, ctx: *mut ffi::V8ContextHandle) -> bool {
        unsafe { ffi::klyron_v8_promise_has_handler(ctx, self.handle) }
    }

    #[cfg(feature = "native")]
    pub fn mark_as_handled(&self, ctx: *mut ffi::V8ContextHandle) -> Result<(), V8Error> {
        let r = unsafe { ffi::klyron_v8_promise_mark_as_handled(ctx, self.handle) };
        if r.success { Ok(()) }
        else { Err(V8Error::CallFailed("mark_as_handled failed".into())) }
    }

    #[cfg(feature = "native")]
    pub fn handle(&self) -> *mut ffi::V8PromiseHandle {
        self.handle
    }
}
